### ISSUE TITLE
[FEATURE]  Expect to support the filesystem not implementing the append-mode.

### ISSUE THREAD

**Issue by yuyang733** [apache/incubator-uniffle#391]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Cloud object storage and its corresponding accelerated cache are widely used in storage-computing separation architecture.  
But many of them may not implement the append-mode or the write amplification of the append-mode will be tricky.

Is it possible to consider supporting the way of independently storing each block as a data file at the same time?

### Motivation

Cloud object storage and its corresponding accelerated cache are widely used in storage-computing separation architecture.  
But many of them may not implement the append-mode or the write amplification of the append-mode will be tricky.

### Describe the solution

The initially envisaged solution is to implement a non-append storage type, such as `COSShuffleWriteHandler`/ `COSShuffleDeleteHandler`, which writes all ShuffleBlocks as a DataFile instead of appending blocks in a dataFile.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by zuston** [comment]
Do you mean that one dataFlushEvent could directly write a new single oss file? If so, how to maintain the index file?

**Comment by yuyang733** [comment]
> Do you mean that one dataFlushEvent could directly write a new single oss file? If so, how to maintain the index file?

Aren't several blocks appended to a data file at present?  I want to write a block into a single data file, so it corresponds to an index file.

**Comment by zuston** [comment]
Could u help give a data layout picture, which includes the relations of data block and oss file name. 

Besides, we should also consider how to support the local_order in object store.

**Comment by yuyang733** [comment]
> Could u help give a data layout picture, which includes the relations of data block and oss file name.
> 
> Besides, we should also consider how to support the local_order in object store.

Okay, I will provide a design gram to describe it later.

**Comment by zuston** [comment]
> Aren't several blocks appended to a data file at present? 

Yes. I know it’s better to avoid using the append mode to improve the performance. But if the data layout is changed, do we need a index file? In the local file or HDFS storage type, the index file will maintain the relation of block offset in the single one file.

**Comment by zuston** [comment]
I have another question that will we directly use the object store api or to use the hadoop filesystem api to support this? 

If it is the former, maybe it’s better to introduce the uniffle dedicated filesystem api to wrap different concrete filesystems, also including the cos.

**Comment by zuston** [comment]
> the index file seems to be not needed anymore.

Emm.. Maybe not. If having no a index structure, it means we don’t a global vision to determine find files we needed in object store, especially for local order. WDYT? @jerqi

**Comment by zuston** [comment]
> Every data file should have index file.

The performance will not be good, as we have to read all index files in one time for doing split segments.

**Comment by jerqi** [comment]
> > Every data file should have index file.
> 
> The performance will not be good, as we have to read all index files in one time for doing split segments.

We can read only one index file.

**Comment by yuyang733** [comment]
> I have another question that will we directly use the object store api or to use the hadoop filesystem api to support this?
> 
> If it is the former, maybe it’s better to introduce the uniffle dedicated filesystem api to wrap different concrete filesystems, also including the cos.

In fact, Can an abstract storage layer be provided in uniffle, and each manufacturer implements the necessary storage interface?

It does not care whether the concrete class uses the native API (such as S3, etc) or the HCFS file system.

**Comment by LuciferYang** [comment]
Thanks for ping me @advancedxy I am not very familiar with object storage, but I think it is better to design it separately to avoid incompatibility of data layout and negative impact on the performance of the current implementation

**Comment by zuston** [comment]
Any update on this? @yuyang733 cc @jerqi @advancedxy 

If object store is supported, I will use this to store huge partition to reduce HDFS pressure for iQiyi. And this is an important feature for uniffle
